automation/orchestrator.go

```diff
@@ -238,92 +306,142 @@ func (o *Orchestrator) RunWorkflow(ctx context.Context, wid workflow.ID) error {
+			OwnerID:    wf.OwnerID,
+			WorkflowID: wf.WorkflowID(),
+			RunID:      runID,
+			Status:     status,
+		}); err != nil {
+			log.Debug(err)
+		}
+	}(wf)
+	return err
 }
 
 // ApplyWorkflow runs the given workflow, but does not record the output
-func (o *Orchestrator) ApplyWorkflow(ctx context.Context, wid workflow.ID) error {
+func (o *Orchestrator) ApplyWorkflow(ctx context.Context, wait bool, scriptOutput io.Writer, wf *workflow.Workflow, ds *dataset.Dataset, secrets map[string]string) (string, error) {
 	o.runLock.Lock()
 	defer o.runLock.Unlock()
-	log.Debugw("ApplyWorkflow, workflow", "id", wid)
+	log.Debugw("ApplyWorkflow, workflow", "id", wf.ID)
 	apply := o.applyFactory(ctx)
-	wf, err := o.GetWorkflow(wid)
-	if err != nil {
-		log.Debugw("ApplyWorkflow: getting workflow from store", "err", err)
-		return fmt.Errorf("getting workflow from store: %w", err)
+
+	runID := run.NewID()
+	if scriptOutput != nil {
+		o.bus.SubscribeID(func(ctx context.Context, e event.Event) error {
+			go func() {
+				log.Debugw("apply transform event", "type", e.Type, "payload", e.Payload)
+				if e.Type == event.ETTransformPrint {
+					if msg, ok := e.Payload.(event.TransformMessage); ok {
+						if scriptOutput != nil {
+							io.WriteString(scriptOutput, msg.Msg)
+							io.WriteString(scriptOutput, "\n")
+						}
+					}
+				}
+			}()
+			return nil
+		}, runID)
+		// TODO (ramfox): defer unsubscribe to id
 	}
-	streams := ioes.NewDiscardIOStreams()
 
 	// TODO (ramfox): when we understand what it means to dryrun a hook, this should wait for the err, iterator thought the hooks
 	// for this workflow, and emit the events for hooks that this orchestrator understands
-	return apply(ctx, streams, wf)
+	return runID, apply(ctx, wait, runID, wf, ds, secrets)
 }
 
-// CreateWorkflow creates a new workflow and adds it to the WorkflowStore
-func (o *Orchestrator) CreateWorkflow(did string, pid profile.ID, triggerOpts []map[string]interface{}) (*workflow.Workflow, error) {
-	t := []trigger.Trigger{}
-	for _, opt := range triggerOpts {
+// SaveWorkflow creates a new workflow if the workflow id is empty, or updates
+// an existing workflow in the workflow Store
+func (o *Orchestrator) SaveWorkflow(wf *workflow.Workflow) (*workflow.Workflow, error) {
+	if wf.ID != "" {
+		fetchedWF, err := o.workflows.Get(wf.ID)
+		if errors.Is(err, workflow.ErrNotFound) {
+			return nil, fmt.Errorf("SaveWorkflow error: workflow %q, %w", wf.ID, err)
+		}
+		if fetchedWF.DatasetID != wf.DatasetID {
+			return nil, fmt.Errorf("SaveWorkflow error: given workflow %q has a different DatasetID than the workflow on record", wf.ID)
+		}
+		if fetchedWF.OwnerID != wf.OwnerID {
+			return nil, fmt.Errorf("SaveWorkflow error: given workflow %q has a different OwnerID than the workflow on record", wf.ID)
+		}
+		if fetchedWF.Created != wf.Created {
+			return nil, fmt.Errorf("SaveWorkflow error: given workflow %q has a different Created time than the workflow on record", wf.ID)
+		}
+	}
```

Comment:
moved the responsibility of catching potential problems out of the `workflow.Store` and into `SaveWorkflow`

Combined `CreateWorkflow` and `UpdateWorkflow` into `SaveWorkflow`